edx_exams/apps/core/tests/test_api.py

```diff
@@ -119,6 +121,89 @@ def test_get_exam_attempt_missing_time_limit(self):
         self.assertEqual(0, time_left)
 
 
+class TestCheckIfExamTimedOut(ExamsAPITestCase):
+    """
+    Tests for API utility function `check_if_exam_timed_out`
+    """
+    # setup:
+
+    def setUp(self):
+        super().setUp()
+
+        self.course_id = 'course-v1:edx+test+f19'
+        self.content_id = 'block-v1:edX+test+2023+type@sequential+block@1111111111'
+
+        self.exam = Exam.objects.create(
+            resource_id=str(uuid.uuid4()),
+            course_id=self.course_id,
+            provider=self.test_provider,
+            content_id=self.content_id,
+            exam_name='test_exam',
+            exam_type='proctored',
+            time_limit_mins=60,
+            due_date='3000-07-01 00:00:00',
+            hide_after_due=False,
+            is_active=True
+        )
+
+        self.one_hour_ago = timezone.now() - timedelta(hours=1)
+
+    def create_mock_attempt(self, status, start_time, allowed_time_limit_mins):
+        return ExamAttempt.objects.create(
+            user=self.user,
+            exam=self.exam,
+            attempt_number=1,
+            status=status,
+            start_time=start_time,
+            allowed_time_limit_mins=allowed_time_limit_mins
+        )
+
+    def test_submit_on_timeout(self):
+        """
+        Test that an attempt id is returned when an in-progress exam times out
+        """
+        exam_attempt_started = self.create_mock_attempt(ExamAttemptStatus.started, self.one_hour_ago, 60)
+        exam_attempt_ready_to_submit = self.create_mock_attempt(ExamAttemptStatus.ready_to_submit, self.one_hour_ago, 60)
+
+        self.assertEqual(check_if_exam_timed_out(exam_attempt_started), 1)
+        self.assertEqual(check_if_exam_timed_out(exam_attempt_ready_to_submit), 2)
+
+    def test_dont_submit(self):
+        """
+        Test that None is returned when an exam is not in-progress or not timed out
+        """
+        # not in progress, but timed out
```

Comment:
This is a great use case for ddt, so you can feed different args to the same test instead of adding them all in the test function itself.

Reply:
Looking through the docs of ddt now... this is the coolest thing ever. Gonna play around with this.

Reply:
It's a super useful tool for testing! Here's a pretty basic example of how it's already being used in edx-exams: https://github.com/edx/edx-exams/blob/613f1577fcd42230a8174499a690f12342e4223f/edx_exams/apps/api/v1/tests/test_views.py#L229-L234